src/HDF5.jl

```diff
@@ -1412,7 +1425,14 @@ function read(obj::HDF5Dataset, ::Union{Type{Array{HDF5Compound}},Type{HDF5Compo
     # Read the raw data
     buf = Vector{UInt8}(length(obj)*sz)
     h5d_read(obj.id, memtype_id, H5S_ALL, H5S_ALL, H5P_DEFAULT, buf)
-    HDF5Compound(buf, membertype, membername, memberoffset)
+
+    # Convert to the appropriate data format using iobuffer
+    iobuff = IOBuffer(buf)
+    data = Any[]
+    while !eof(iobuff)
+        push!(data, read_row(iobuff, membertype, membersize))
+    end
+    HDF5Compound(data, membername)
```

Comment:
I think the return type should depend upon whether the input storage format is `Array{HDF5Compound}` or just `HDF5Compound`, where we return `Vector{HDF5Compound}` in the former case and a single `HDF5Compound` element in the latter.